File: framework/web/src/main/java/top/bettercode/simpleframework/web/serializer/ArraySerializer.java
package top.bettercode.simpleframework.web.serializer;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonStreamContext;
import com.fasterxml.jackson.databind.BeanProperty;
import com.fasterxml.jackson.databind.JsonSerializer;
import com.fasterxml.jackson.databind.SerializerProvider;
import com.fasterxml.jackson.databind.annotation.JacksonStdImpl;
import com.fasterxml.jackson.databind.jsontype.TypeSerializer;
import com.fasterxml.jackson.databind.ser.ContextualSerializer;
import com.fasterxml.jackson.databind.ser.std.StdScalarSerializer;
import java.io.IOException;
import org.springframework.util.StringUtils;
import top.bettercode.simpleframework.web.serializer.annotation.JsonArray;

/**
 * JSON序列化url自动补全
 *
 * @author Peter Wu
 */
@JacksonStdImpl
public class ArraySerializer extends StdScalarSerializer<String> implements
    ContextualSerializer {

  private static final long serialVersionUID = 1L;

  private final boolean useExtensionField;
  private final String separator;

  public ArraySerializer() {
    this(true, ",");
  }

  public ArraySerializer(
      boolean useExtensionField,
      String separator) {
    super(String.class, false);
    this.useExtensionField = useExtensionField;
    this.separator = separator;
  }


  @Override
  public void serialize(String value, JsonGenerator gen, SerializerProvider provider)
      throws IOException {
    if (useExtensionField) {
      gen.writeString(value);
      JsonStreamContext outputContext = gen.getOutputContext();
      String fieldName = outputContext.getCurrentName();
      gen.writeObjectField(fieldName + "Array",
          (StringUtils.hasText(value) ? value.split(separator) : new String[0]));
    } else {
      gen.writeObject(
          (StringUtils.hasText(value) ? value.split(separator) : new String[0]));
    }
  }


  @Override
  public final void serializeWithType(String value, JsonGenerator gen, SerializerProvider provider,
      TypeSerializer typeSer) throws IOException {
    serialize(value, gen, provider);
  }

  @Override
  public JsonSerializer<?> createContextual(SerializerProvider prov, BeanProperty property) {
    if (property != null) {
      JsonArray annotation = property.getAnnotation(JsonArray.class);
      if (annotation == null) {
        throw new RuntimeException("未注解@" + JsonArray.class.getName());
      }
      return new ArraySerializer(annotation.extended(), annotation.value());
    }
    return this;
  }
}

File: web/src/main/kotlin/top/bettercode/summer/web/serializer/ArraySerializer.kt
package top.bettercode.summer.web.serializer

import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.databind.BeanProperty
import com.fasterxml.jackson.databind.JsonSerializer
import com.fasterxml.jackson.databind.SerializerProvider
import com.fasterxml.jackson.databind.annotation.JacksonStdImpl
import com.fasterxml.jackson.databind.jsontype.TypeSerializer
import com.fasterxml.jackson.databind.ser.ContextualSerializer
import com.fasterxml.jackson.databind.ser.std.StdScalarSerializer
import org.springframework.util.StringUtils
import top.bettercode.summer.web.serializer.annotation.JsonArray

/**
 * JSON序列化url自动补全
 *
 * @author Peter Wu
 */
@JacksonStdImpl
class ArraySerializer @JvmOverloads constructor(
        private val useExtensionField: Boolean = true,
        private val separator: String = ",") : StdScalarSerializer<String>(String::class.java, false), ContextualSerializer {
    override fun serialize(value: String, gen: JsonGenerator, provider: SerializerProvider) {
        if (useExtensionField) {
            gen.writeString(value)
            val outputContext = gen.outputContext
            val fieldName = outputContext.currentName
            gen.writeObjectField(fieldName + "Array",
                    if (StringUtils.hasText(value)) value.split(separator.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray() else arrayOf(0))
        } else {
            gen.writeObject(
                    if (StringUtils.hasText(value)) value.split(separator.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray() else arrayOf(0))
        }
    }

    override fun serializeWithType(value: String, gen: JsonGenerator, provider: SerializerProvider,
                                   typeSer: TypeSerializer) {
        serialize(value, gen, provider)
    }

    override fun createContextual(prov: SerializerProvider, property: BeanProperty?): JsonSerializer<*> {
        if (property != null) {
            val annotation = property.getAnnotation(JsonArray::class.java)
                    ?: throw RuntimeException("未注解@" + JsonArray::class.java.name)
            return ArraySerializer(annotation.extended, annotation.value)
        }
        return this
    }

}
